plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -212,9 +223,50 @@ public void asyncBlobUpload(WriteContext writeContext, ActionListener<Void> comp
         }
     }
 
+    @ExperimentalApi
     @Override
     public void readBlobAsync(String blobName, ActionListener<ReadContext> listener) {
-        throw new UnsupportedOperationException();
+        try (AmazonAsyncS3Reference amazonS3Reference = SocketAccess.doPrivileged(blobStore::asyncClientReference)) {
+            final S3AsyncClient s3AsyncClient = amazonS3Reference.get().client();
+            final String bucketName = blobStore.bucket();
+            final AsyncTransferManager transferManager = blobStore.getAsyncTransferManager();
+
+            final GetObjectAttributesResponse blobMetadata = getBlobMetadata(s3AsyncClient, bucketName, blobName).get();
+
+            final long blobSize = blobMetadata.objectSize();
+            final int numberOfParts = blobMetadata.objectParts().totalPartsCount();
+            final String blobChecksum = blobMetadata.checksum().checksumCRC32();
+
+            final List<InputStreamContainer> blobPartStreams = new ArrayList<>();
```

Comment:
You can get rid of this intermediate list and the extra future handling by doing the following. What do you think?

```java
final List<CompletableFuture<InputStreamContainer>> blobPartInputStreamFutures = new ArrayList<>();
// S3 multipart files use 1 to n indexing
for (int partNumber = 1; partNumber <= numberOfParts; partNumber++) {
    blobPartInputStreamFutures.add(getBlobPartInputStreamContainer(
        s3AsyncClient,
        bucketName,
        blobName,
        partNumber
    ));
}

CompletableFuture.allOf(blobPartInputStreamFutures.toArray(CompletableFuture[]::new)).whenComplete((unused, throwable) -> {
    if (throwable == null) {
        listener.onResponse(new ReadContext(blobSize,
            blobPartInputStreamFutures.stream()
                .map(CompletableFuture::join)
                .collect(Collectors.toList()),
            blobChecksum));
    } else {
        Exception ex = throwable instanceof Exception ? (Exception) throwable.getCause() : new Exception(throwable.getCause());
        listener.onFailure(ex);
    }
});
```